etcdserver: avoid deadlock caused by adding members with wrong peer URLs

mitake opened the pull request:
Current membership changing functionality of etcd seems to have a
problem which can cause deadlock.

How to produce:
 1. construct N node cluster
 2. add N new nodes with etcdctl member add, without starting the new members

What happens:
After finishing add N nodes, a total number of the cluster becomes 2 *
N and a quorum number of the cluster becomes N + 1. It means
membership change requires at least N + 1 nodes because Raft treats
membership information in its log like other ordinal log append
requests.

Assume the peer URLs of the added nodes are wrong because of miss
operation or bugs in wrapping program which launch etcd. In such a
case, both of adding and removing members are impossible because the
quorum isn't preserved. Of course ordinal requests cannot be
served. The cluster would seem to be deadlock.

Of course, the best practice of adding new nodes is adding one node
and let the node start one by one. However, the effect of this problem
is so serious. I think preventing the problem forcibly would be
valuable.

Solution:
This patch lets etcd forbid adding a new node if the operation changes
quorum and the number of changed quorum is larger than a number of
running nodes.

Fixes https://github.com/coreos/etcd/issues/3477

xiang90 commented:
@mitake This is what we wanted to do for a while. We will review and get this merged after 2.2 release today. Thanks!

mitake commented:
@xiang90 , thanks, I'm looking forward your review :)

BTW, how do you think about an offline tool (etcdck?) for handling this sort of problem e.g. check and manipulate membership forcibly. If this idea is acceptable, I'd like to work on it.

AdoHe commented:
@mitake @xiang90 this seems a little bit violent.

yichengq commented:
@AdoHe > this seems a little bit violent.

What is your thoughts about it? I think the idea is reasonable: no one would like to have a cluster which loses quorum and is unrecoverable.

xiang90 commented:
@AdoHe This is actually pretty conservative. These member are not even started. So it is safe.

xiang90 commented on etcdserver/cluster.go:
```diff
@@ -253,6 +253,23 @@ func (c *cluster) ValidateConfigurationChange(cc raftpb.ConfChange) error {
 			}
 		}
+
+		nrRunningNodes := 0
```
can we move this code block to a func called `checkLiveness` (maybe a better func name)?

xiang90 replied:
we do not use `node`. we use `member` instead.

mitake replied:
How about countRunningMembers()?

mitake replied:
countStartedMembers() would be suitable, I think.

xiang90 commented on etcdserver/cluster.go:
```diff
@@ -253,6 +253,23 @@ func (c *cluster) ValidateConfigurationChange(cc raftpb.ConfChange) error {
 		}
+
+		nrRunningNodes := 0
+		nrAllNodes := 0
```
how about initializing this to 1? then we do not need line 266

xiang90 commented on etcdserver/cluster.go:
```diff
@@ -253,6 +253,23 @@ func (c *cluster) ValidateConfigurationChange(cc raftpb.ConfChange) error {
+		quorum := nrAllNodes/2 + 1
+		plog.Infof("nrAllNodes: %d, nrRunningNodes: %d, quorum: %d", nrAllNodes, nrRunningNodes, quorum)
+		if nrRunningNodes < quorum {
+			plog.Errorf("a number of running nodes (%d) will be less than a quorum number of the cluster (%d)", nrRunningNodes, quorum)
```
this should be a warning instead of error.

xiang90 commented on etcdserver/errors.go:
```diff
@@ -31,6 +31,7 @@ var (
 	ErrTimeout                    = errors.New("etcdserver: request timed out")
 	ErrTimeoutDueToLeaderFail     = errors.New("etcdserver: request timed out, possibly due to previous leader failure")
 	ErrTimeoutDueToConnectionLost = errors.New("etcdserver: request timed out, possibly due to connection lost")
+	ErrNotEnoughRunningNodes      = errors.New("etcdserver: cluster doesn't have enough running nodes")
```
ErrNotEnoughStartedMembers

xiang90 replied:
cluster does not have enough started members

xiang90 commented on etcdserver/cluster.go:
```diff
@@ -253,6 +253,23 @@ func (c *cluster) ValidateConfigurationChange(cc raftpb.ConfChange) error {
+		}
+
+		nrAllNodes++ // this increment is for the newly added node
+		quorum := nrAllNodes/2 + 1
```
nquorum = nmembers/2 + 1

AdoHe commented:
@xiang90 @yichengq I agree with what @xiang90  commented, we should move this code block to `checkLiveness` func, this makes more sense. For operations which change quorum if liveness is down, it should be forbidden.

mitake commented:
@xiang90 , thanks for your review. I updated this PR based on your comments.

xiang90 commented on etcdserver/cluster.go:
```diff
@@ -253,6 +265,16 @@ func (c *cluster) ValidateConfigurationChange(cc raftpb.ConfChange) error {
 			}
 		}
+
+		nmembers := len(c.members) + 1
```
i should be more clear: i hope we can move all of the new code to a new func.

xiang90 commented:
@mitake After discussing with @yichengq, we think this has to be done before the command is submitted into raft. This is for backward compatibility reason. Or by running the same set of commands, etcd 2.1 and etcd 2.2 might have different behavior after this pr.

mitake commented:
@xiang90 @yichengq Should this check be done in etcdctl?

BTW, is rolling update from 2.1 -> 2.2 supported? If so, I agree with not doing the check in server side. However, if the rolling update isn't supported, the check wouldn't be harmful.

In addition, seems that v2.2.0 is already released. Do you have a plan of backporting this patch to the stable releases? It also affects the compatibility problem.

xiang90 commented:
@mitake We do support rolling upgrade. I need to think though this more.

mitake commented:
@xiang90 How about check in both of server and clients (etcdctl and client/ go library)? The effect of this problem is critical, so checking in all components would be useful for defensive purpose. In addition, you can backport client side change to stable releases and prevent the problem without breaking rolling upgrade.

xiang90 commented:
@mitake The problem is that we do not control all clients/ctls. We only control go client. We may add this checking into server side. However, the checking has to be done before sending the reconfiguration into raft.

mitake commented:
@xiang90 So do you agree with checking in both of server and etcd official clients?

xiang90 commented:
@mitake Let do it step by step. I think we can do it at server side before sending the command into raft.

mitake commented:
@xiang90 ok, I agree with doing it step by step. So what I need to do in this PR is cleaning the code of calculating nmembers, right?

mitake commented:
@xiang90 @yichengq updated this PR for unifying changes into single function.

mitake commented:
@xiang90 ah, sorry, seems that I misunderstand your intention. Should I insert the check in outside of raft logic? e.g. EtcdServer.AddMember()?

yichengq commented:
@mitake i think so. This helps to avoid raft-log replay mismatch in 2.1 and 2.2. If you add the check when applying log, it may happen that 2.2 rejects some add command while 2.1 executes it.

Another problem is that etcd needs to support adding 1 member to 1-member cluster because this is a necessary step in disaster recovery today.

mitake commented:
@yichengq I understood. However, I think adding the check to raft side would be better for handling a case of member removal. The deadlock problem can happen even in the removal case like this:

1. a cluster has N + 1 started nodes and N unstarted nodes with wrong peer URLs, quorum is N + 1
2. user removes one started node: quorum remains N + 1
3. the deadlock can happen because a number of started node is less than the quorum

I think serializing the checking in raft state machine can prevent this problem. If we do it in requester's side, two removal request can break quorum potentially.

Am I too pessimistic? :) Of course I'd like to esteem maintainers' opinion.

BTW,
> adding 1 member to 1-member cluster
what does it mean? I couldn't understand "1-member".

mitake commented:
@yichengq BTW, is it impossible to avoid backporting this change to stable releases?
I know nothing about the release style of etcd, sorry :(

yichengq commented:
@mitake We have to support rolling upgrade, which means that etcd cluster should ALWAYS be able to upgrade from 2.2 to 2.3. Assuming this PR is merged in 2.3, if we add check in apply process, it may break the upgrade path, and that is bad. That is the reason that avoid backporting doesn't solve this problem.

The idea about removing member is another thing, and it is not related to this PR. Let us do it step by step.

I refer 1-member cluster as the cluster that has only 1 member. Check this cluster restore procedure (https://github.com/coreos/etcd/blob/master/Documentation/admin_guide.md#restoring-the-cluster), and you could find it needs to support changing cluster size from 1 to 2.

mitake commented:
@yichengq I missed a case of 2.2 -> 2.3, I understand. I'll move the checking to etcdserver side. I'll consider about the removal case in the future.

And thanks for your explanation about 1-member cluster!

mitake commented:
@yichengq @xiang90 updated this PR, changed points:
- move the checking logic to EtcdServer
- handle a 1-member cluster case

yichengq commented:
@mitake Good move. It no longer affects upgrade correctness now.

One thing is that you need to fix the broken test. You could see `continuous-integration/travis-ci/pr` is failing.

Another thing is that this is a HTTP API change, and it may break some people still. For example, users may do automatic tests on etcd and check its behavior when keep adding members. Adding a `-strict-check` flag could be a good choice. /cc @xiang90

mitake commented:
@yichengq ah, I'll fix the broken tests.

How about `-strict-membership-check` as a name of the option?

yichengq commented:
`-strict-reconfig-check` sounds better to me because it is more specific.

yichengq commented on etcdserver/server.go:
```diff
@@ -602,7 +602,41 @@ func (s *EtcdServer) LeaderStats() []byte {
+	nstarted := 0
+
+	for _, member := range members {
+		if len(member.Name) != 0 {
```
One more tip: i would like to make this a member function because it is implicit that `len(member.Name) != 0` -> unstarted member`

mitake replied:
ok, i'll make it a member function in the next verison

mitake commented:
-strict-reconfig-check seems good, i'll use the name

mitake commented:
I updated this PR.

- Added a new option `-strict-reconfig-check` to etcd. If this option isn't passed to etcd, it will keep the existing behavior.
- isReadyToAddNewMember() is now a member function of EtcdServer

mitake commented:
I forgot to update the test. I'll fix it soon.

yichengq commented on etcdmain/config.go:
```diff
@@ -177,6 +178,7 @@ func NewConfig() *config {
 		// Should never happen.
 		plog.Panicf("unexpected error setting up clusterStateFlag: %v", err)
 	}
+	fs.BoolVar(&cfg.strictReconfigCheck, "strict-reconfig-check", false, "Reject member change if a number of running node is less than quorum for avoid deadlock in a case of wrong peer URLs")
```
Reject to adding member if the number of started nodes is less than quorum.

I would not talk about specific use case of this check here. As long as our users could know what check does, they could understand what it is for. And the check could reject other unreasonable use cases.

mitake replied:
OK, how about this: Reject membership change if the change can cause potential deadlock

yichengq replied:
This is fine to me. s/membership/reconfiguration to match the flag.

mitake replied:
I got it

yichengq commented on etcdserver/server.go:
```diff
@@ -602,7 +602,47 @@ func (s *EtcdServer) LeaderStats() []byte {
+	nstarted := 0
+
+	for _, member := range s.cluster.members {
+		if len(member.Name) != 0 {
```
You forgot to make this a member function..

mitake replied:
Ah, sorry, I missed your intention. I'll add a member function for checking started or not to Member in the next update.

BTW, I made isReadyToAddNewMember() a member function. Is this ok?

yichengq replied:
I suppose you mean `a EtcdServer function`. It is better if you could make this a helper function, which takes `cluster` as its parameter.

yichengq commented on etcdserver/server.go:
```diff
@@ -602,7 +602,47 @@ func (s *EtcdServer) LeaderStats() []byte {
+	}
+
+	nquorum := nmembers/2 + 1
+	plog.Infof("all nodes: %d (quorum: %d), started nodes: %d", nmembers, nquorum, nstarted)
```
I prefer the line could express the idea that the add request passes check, but it is not guaranteed to be committed and applied.

yichengq replied:
Or you could just drop this line.

mitake replied:
Do you mean the log of line 631 is simply needless?

yichengq replied:
We have printed the add member log line here: https://github.com/coreos/etcd/blob/master/etcdserver/server.go#L867

It is ok for me to log that it passes strict check. You could judge it by yourself: Run this workflow and read through all logs to see whether it makes sense.

yichengq commented:
It has reached the state that I am almost OK with it. :+1:

mitake commented:
@yichengq updated based on your comments, thanks for review!

mitake commented:
@yichengq updated again just for removing the log of etcdserver/server.go line 631

yichengq commented:
@mitake How do you like the idea to make `isReadyToAddNewMember` a function of `cluster struct`? It only uses information in `cluster`.

xiang90 commented on etcdmain/config.go:
```diff
@@ -177,6 +178,7 @@ func NewConfig() *config {
 		// Should never happen.
 		plog.Panicf("unexpected error setting up clusterStateFlag: %v", err)
 	}
+	fs.BoolVar(&cfg.strictReconfigCheck, "strict-reconfig-check", false, "Reject reconfiguration if it can cause potential deadlock")
```
I guess we need to improve the message "Reject reconfiguration if it can cause potential deadlock".

Potential deadlock is not that obvious to a lot users. 

"Reject reconfiguration that might cause quorum loss"

xiang90 commented on etcdserver/server.go:
```diff
@@ -602,7 +602,45 @@ func (s *EtcdServer) LeaderStats() []byte {
+		// a case of adding a new node to 1-member cluster for restoring cluster data
+		// https://github.com/coreos/etcd/blob/master/Documentation/admin_guide.md#restoring-the-cluster
+
+		plog.Infof("a number of started nodes is 1, this cluster is ready to add a new node")
```
let's move this to debug level. Also `The number of started member is 1. This cluster can accept add member request.`

xiang90 commented on etcdserver/server.go:
```diff
@@ -602,7 +602,45 @@ func (s *EtcdServer) LeaderStats() []byte {
+
+	nquorum := nmembers/2 + 1
+	if nstarted < nquorum {
+		plog.Warningf("a number of started nodes (%d) will be less than a quorum number of the cluster (%d)", nstarted, nquorum)
```
Reject add member request: the number of started member (%d) will be less than the quorum number of the cluster (%d)

mitake commented:
@yichengq `isReadyToAddNewMember` depends on EtcdServer.cfg. I think the function should belong to EtcdServer for now.

mitake commented:
updated the messages based on @xiang90 's comments

xiang90 commented:
LGTM. Defer to @yichengq

xiang90 commented:
@mitake Thanks so much for your patient and contribution!

jonboulle commented on etcdserver/errors.go:
```diff
@@ -31,6 +31,7 @@ var (
 	ErrTimeout                    = errors.New("etcdserver: request timed out")
 	ErrTimeoutDueToLeaderFail     = errors.New("etcdserver: request timed out, possibly due to previous leader failure")
 	ErrTimeoutDueToConnectionLost = errors.New("etcdserver: request timed out, possibly due to connection lost")
+	ErrNotEnoughStartedMembers    = errors.New("etcdserver: re-configuration faild due to not enough started members")
```
failed

mitake replied:
thanks, I'll fix it.

jonboulle commented on etcdserver/server.go:
```diff
@@ -602,7 +602,45 @@ func (s *EtcdServer) LeaderStats() []byte {
 
 func (s *EtcdServer) StoreStats() []byte { return s.store.JsonStats() }
 
+func (s *EtcdServer) isReadyToAddNewMember() bool {
```
Unit test for this logic would be great.

mitake replied:
ok, I'll add a unit test

mitake commented:
@yichengq on the second thought, moving `isReadyToAddNewMember` to `cluster struct` would be a good idea. If we call the func from `EtcdServer.AddMember` and add a condition on cfg to `AddMember`, it only depends on `cluster`. I'll fix it.

mitake commented:
updated, just moved `isReadyToAddNewMember` to `cluster struct`

mitake commented:
@jonboulle should I add a unit test and update this PR soon? If so, it would take for a while because I'm new to testing in go, sorry :(

jonboulle commented:
@mitake I think this should have a test before it is merged. To help you out I've started a test table here for you: https://github.com/jonboulle/etcd/commit/9c3c6650bc7258204276d085e571aefb159d48e9

While writing the tests I also discovered that https://github.com/coreos/etcd/pull/3479/files#diff-62824440a438e1c7efbb9cf76299624eR382 should also be checking the cluster size, so as you can see I updated that line too.

Can you cherry-pick my commit onto your branch and then perhaps add some more cases as you see fit? thanks!

yichengq commented:
The code is totally awesome now.

@jonboulle Thanks for helping on unit test. ;)

@mitake As jon suggests, it is cool if you could add some unit test. Because we have separate the check logic in a single function, it is pretty simple to add a unit test for the new logic. Table testing is a common test method used in Go, and it could cover multiple test cases in an easy-to-read way.

mitake commented:
@jonboulle thanks a lot for your help! I'll enhance your test soon.

@yichengq I'll update this PR after enhancing tests.

mitake commented:
@jonboulle @yichengq enhanced test for cases of even members (want success) and empty cluster

mitake commented:
updated for gofmt style problem

yichengq commented:
LGTM.

Thanks for the contribution! The PR is quite fantastic now.

mitake commented:
Thanks a lot for your review and help, @xiang90 @yichengq @jonboulle !
